autosklearn/ensemble_builder.py

```diff
@@ -210,39 +427,37 @@ def __init__(
         self.datamanager = self.backend.load_datamanager()
         self.y_valid = self.datamanager.data.get('Y_valid')
         self.y_test = self.datamanager.data.get('Y_test')
+        self.ensemble_history = []
 
-        # Support for tracking the performance across time
-        # A Queue is needed to handle multiprocessing, not only
-        # internally for pynisher calls, but to return data
-        # to the main process
-        # Hence, because we are using three different processes,
-        # the below strategy prevents MemoryErrors. That is,
-        # without clearly isolating the queue with a manger,
-        # we run into a threading MemoryError
-        if queue is None:
-            mgr = multiprocessing.Manager()
-            mgr.Namespace()
-            self.queue = mgr.Queue()
-        else:
-            self.queue = queue
-        self.queue.put([])
-        self.queue.get()
+    @classmethod
+    def _get_ensemble_logger(self, logger_name, dirname):
+        """
+        Returns the logger of for the ensemble process.
+        A subprocess will require to set this up, for instance,
+        pynisher forks
+        """
+        setup_logger(
+            os.path.join(
+                dirname,
+                '%s.log' % str(logger_name)
+            ),
+        )
 
-    def run(self):
-        buffer_time = 5  # TODO: Buffer time should also be used in main!?
+        return get_logger('EnsembleBuilder')
+
+    def run(self, time_left, iteration, time_buffer=5):
         process_start_time = time.time()
         while True:
             time_elapsed = time.time() - process_start_time
-            time_left = self.time_limit - buffer_time - time_elapsed
+            time_left -= time_elapsed
             if time_left < 1:
                 break
-            self.time_left = time_left
             safe_ensemble_script = pynisher.enforce_limits(
-                wall_time_in_s=int(time_left),
+                wall_time_in_s=int(time_left + time_buffer),
                 mem_in_mb=self.memory_limit,
                 logger=self.logger
             )(self.main)
-            safe_ensemble_script()
+            safe_ensemble_script(time_left, iteration)
```

Comment:
I think the safeguards below which half the memory usage will no longer work together with the new code structure. As the ensemble builder is re-constructed in every iteration, it can't remember the value for nbest and this needs to be done above in the for loop.

Reply:
Hi Matthias, the nbest is remembered through this line:

https://github.com/automl/auto-sklearn/blob/8fc9d3090887d0bb5383000f1e1cb72757f1a38d/autosklearn/ensemble_builder.py#L213

Reply:
I see. Do we have a test for this? Also, do we delete temporary file that might be too large now?

Also, what's the else in the return statement returning exactly and when does it happen?